scheduler/src/cook/queue_limit.clj

```diff
@@ -0,0 +1,155 @@
+(ns cook.queue-limit
+  (:require [clj-time.core :as time]
+            [clojure.tools.logging :as log]
+            [cook.caches :as caches]
+            [cook.config :as config]
+            [cook.datomic :as datomic]
+            [cook.queries :as queries]
+            [cook.regexp-tools :as regexp-tools]
+            [cook.util :as util]
+            [datomic.api :as d]
+            [plumbing.core :as pc]
+            [metrics.timers :as timers]))
+
+(defn- match-based-on-pool-name
+  "Given a pool name and field name, returns
+  the value for the first matching configuration
+  entry, or Integer/MAX_VALUE if there is no match"
+  [pool-name field]
+  (or
+    (regexp-tools/match-based-on-pool-name
+      (:per-pool (config/queue-limits))
+      pool-name
+      field)
+    Integer/MAX_VALUE))
+
+(defn- pool-global-threshold
+  "Returns the pool-global-threshold for the given pool,
+  the value at which the per-user queue limit switches
+  from the 'normal' number to the 'constrained' number"
+  [pool-name]
+  (match-based-on-pool-name
+    pool-name
+    :pool-global-threshold))
+
+(defn- user-limit-normal
+  "Returns the user-limit-normal for the given pool"
+  [pool-name]
+  (match-based-on-pool-name
+    pool-name
+    :user-limit-normal))
+
+(defn- user-limit-constrained
+  "Returns the user-limit-constrained for the given pool"
+  [pool-name]
+  (match-based-on-pool-name
+    pool-name
+    :user-limit-constrained))
+
+(defn- update-interval-seconds
+  "Returns the interval in seconds at which
+  to refresh queue lengths from the database"
+  []
+  (:update-interval-seconds
+    (config/queue-limits)))
+
+(defn get-pending-jobs
+  "Queries for and returns the set of
+  currently pending jobs from the database"
+  []
+  (-> datomic/conn
+      d/db
+      queries/get-pending-job-ents))
+
+(defn query-queue-lengths
+  "Queries for pending jobs from the database and
+  returns a map with two sub-maps of the following shape:
+
+  {:pool->queue-length {pool-a 100 pool-b 200 ...}
+   :pool->user->queue-length {pool-a {user-x 10 user-y 20 user-z 70}
+                              pool-b {user-x 20 user-y 40 user-z 140}
+                              ...}"
+  []
+  (let [pending-jobs
+        (get-pending-jobs)
+        pool->pending-jobs
+        (group-by
+          caches/job->pool-name
+          pending-jobs)]
+    {:pool->queue-length
```

Comment:
Small optimization: If you do the per pool per-user map first, then you can reduce that to compute the overall length per pool.

Reply:
Done